### PR TITLE
Allow writing volume data at single precision

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Most 3d simulations aren't even accurate to single precision, so writing the volume data at single precision just saves a factor of 2 in disk space usage. With the changes here, users can control whether to write volume data at double or single precision.

Enabling HDF5 compression also helps reduce file size but there's some importer code that uses otherwise untested features that seems to break that I need to fix before I can submit a PR for volume data compression.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/IO/Test_H5.cpp
Comment (kidder) [inline tests/Unit/IO/Test_H5.cpp]: can you check with a number that is not equivalent in float vs double (e.g. 1./3.)?

File: src/DataStructures/Tensor/TensorData.cpp
Comment (kidder) [inline src/DataStructures/Tensor/TensorData.cpp]: space between ) and in

File: src/Evolution/DgSubcell/Events/ObserveFields.hpp
Comment (kidder) [inline src/Evolution/DgSubcell/Events/ObserveFields.hpp]: too many onces

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (kidder) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: again too many onces

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (kidder) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: again

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
I've rebased and pushed fixups. Thanks for the review :)

File: cmake/SpectreSetupPythonPackage.cmake
Comment (nilsvu) [inline cmake/SpectreSetupPythonPackage.cmake]: drop this commit (it's already in develop)

File: tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
Comment (nilsvu) [inline tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml]: I suggest these user-interface changes to make the input files friendlier for users:

- Name `FloatingPointType` -> `Precision` or `FloatingPointPrecision`, case `Float` -> `Single`
- Group the two options in this event, e.g. name the group `Precision` with options `Coordinates` and `Fields`
- Make the type for the fields precision a `std::variant` so it's either a single value or a vector with the correct size.

The result could look like this:

```yaml
Precision:
  Coordinates: Single
  Fields: Double
  # or:
  Fields: [Single, Double, Single, ...]
```

**Review by nilsvu — CHANGES_REQUESTED** [review]
Not a full review yet. This is a great feature though, thanks for doing this!

File: tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
Comment (nilsdeppe) [reply tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml]: I have nothing against the proposed names, though I can't really allocate the time to changing the names since updating them requires changing a bunch of input files.

File: src/IO/H5/VolumeData.cpp
Comment (fmahebert) [inline src/IO/H5/VolumeData.cpp]: Missing space between ")in"

File: src/IO/H5/VolumeData.cpp
Comment (fmahebert) [inline src/IO/H5/VolumeData.cpp:224]: [nitpick] Suggest a blank line above for legibility

File: tests/InputFiles/Elasticity/BentBeam.yaml
Comment (fmahebert) [inline tests/InputFiles/Elasticity/BentBeam.yaml]: I'd suggest the simpler syntax `[Double]` since they're all set to the same type

The file `PlaneWave1DObserveExample.yaml` does a good job of showing the general syntax, so I don't think we need to everywhere.

File: tests/InputFiles/Elasticity/HalfSpaceMirror.yaml
Comment (fmahebert) [inline tests/InputFiles/Elasticity/HalfSpaceMirror.yaml]: again

File: tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
Comment (fmahebert) [inline tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml]: For my understanding: why are you changing the behavior in this input file (and in the Xcts/Schwarzschild file)?

**Review by fmahebert — COMMENTED** [review]
Not a full review, but a quick pass because I was curious. I like this feature!

File: tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml
Comment (nilsdeppe) [reply tests/InputFiles/GeneralizedHarmonic/KerrSchild.yaml]: probably some copy-paste thing or something. Changing to `double`

**Comment by nilsdeppe** [comment]
I've rebased on develop, dropped the python commit, and pushed fixups for @fmahebert's suggestions. Thanks for the (partial) reviews :)

**Comment by fmahebert** [comment]
I'm happy with the fixups to my review.

Also, there's no need to wait for an approval from me to merge this.

**Review by kidder — COMMENTED** [review]
you have my permission to squash

**Comment by nilsdeppe** [comment]
I've rebased on develop and squashed. Thanks for the reviews :)

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@nilsleiffischer please look at

File: src/Evolution/DgSubcell/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/Evolution/DgSubcell/Events/ObserveFields.hpp]: perhaps just grab this and the tag above from `::dg::Events::ObserveFields` (with a type alias that you can also use for the `call_operator_impl`)?

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp:132]: Re https://github.com/sxs-collaboration/spectre/pull/3239#discussion_r646040722:

Here's a suggestion to make the user-interface friendlier with minimal changes:

1. Give this tag and the one below a `std::string name()` ("Fields" and "Coordinates") and put them in an option group named "Precision".
2. In `operator<<` for `FloatingPointType` change "Float" -> "Single"
3. Update the input files to:
    ```yaml
    Precision:
      Coordinates: Single
      Fields: [Double]
    ```

This doesn't require any time-consuming renames in the code but hopefully makes the input files easier to understand for users.

Comment (nilsdeppe) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:132]: It's the input file changes that are the most tedious... There are 9 input files currently that need to be updated with different indentation, etc. so basically _any_ rename is the same amount of work...

File: src/Evolution/DgSubcell/Events/ObserveFields.hpp
Comment (nilsdeppe) [reply src/Evolution/DgSubcell/Events/ObserveFields.hpp]: I've rebased and pushed a fixup for this. Saves quite a few lines of code, thanks for the suggestion :)

**Comment by kidder** [comment]
you have my permission to squash

**Comment by kidder** [comment]
@nilsleiffischer please look at

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:132]: Since copy-pasting the yaml snippet above into 9 input files takes a few minutes at most (10 minutes max including the code changes) and hopefully makes the input files easier to understand for users, my suggestion is to just make this change..

Comment (nilsdeppe) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:132]: Thanks for the suggestion! I can't allocate the time to making this change, I spent a decent amount of time rebasing and so can't spend more time on it.

**Comment by kidder** [comment]
Let's just get this in.  Please squash, I will do the input file changes in a follow-up PR.

**Review by kidder — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
rebased again to deal with conflicts.

**Comment by nilsdeppe** [comment]
Rebased on develop to get the CI fixes

**Review by kidder — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]

**Pull request merged**
